core/query/temporalQuery.ml

```diff
@@ -0,0 +1,857 @@
+open Utility
+open CommonTypes
+open Types
+
+(* Query compilation for temporal queries. *)
+
+module Q = QueryLang
+let base = Q.base
+
+module OpHelpers = struct
+  let binop op x y = Sql.Apply (op, [x; y])
+  let lt = binop "<"
+  let lte = binop "<="
+  let gt = binop ">"
+  let gte = binop ">="
+  let _eq = binop "=="
+  let op_and = binop "&&"
+  let _op_or = binop "||"
+end
+
+(* Shared between valid and transaction time *)
+let current_insertion table_name field_names from_field to_field rows =
+  let compile_rows =
+      List.map (List.map (Q.expression_of_base_value ->- base [])) in
+
+  let rows = compile_rows rows in
+  let field_names = field_names @ [from_field; to_field] in
+  let now = Sql.Constant (Constant.DateTime.now ()) in
+  let forever = Sql.Constant (Constant.DateTime.forever) in
+  let rows = List.map (fun vs -> vs @ [now; forever]) rows in
+  Sql.(Insert {
+    ins_table = table_name;
+    ins_fields = field_names;
+    ins_records = Values rows })
+
+
+module TransactionTime = struct
+
+  let insert = current_insertion
+
+  let update :
+    Types.datatype StringMap.t ->
+    ((Ir.var * string) * Q.t option * Q.t) ->
+    string ->
+    string ->
+    Sql.query =
+  fun table_types ((tbl_var, table), where, body) tt_from tt_to ->
+
+    let now_const = Constant.DateTime.now () in
+    let sql_now = Sql.Constant now_const in
+    let forever_const = Constant.DateTime.forever in
+    let sql_forever = Sql.Constant forever_const in
```

Comment:
Could be hoisted since similar constants are used elsewhere

Reply:
Agreed on `forever`, but given that `now` is side-effecting, hoisting that has little benefit.